File: apps/Clock/Stopwatch.py
from LCD_Lib import LCD
import time
from keys import keyA, keyB, keyY#, keyX

ms_passed = 0

def displaytime(ms_passed):
    hours = ms_passed // 3600000
    minutes = (ms_passed % 3600000) // 60000
    seconds = (ms_passed % 60000) // 1000
    ms = ms_passed % 1000
    stopwatch = "%02d" % int(hours) + ":" + "%02d" % int(minutes) + ":" + "%02d" % int(seconds) + "." + "%03d" % int(ms)
    LCD.fill(0x0000)
    LCD.text("Stopwatch", 40, 40, LCD.green, 2)
    LCD.text(stopwatch, 20, 120, LCD.white, 2)
    LCD.show()

def load_in():
    LCD.fill(0x0000)
    LCD.text("Stopwatch", 60, 40, LCD.green)
    displaytime(0)
    LCD.show()

def startstopwatch():
    running = True
    offset = 0
    start = time.ticks_ms()
    while True:
        if running == True:
            ms_passed = (time.ticks_diff(time.ticks_ms(), start)) + offset
        displaytime(ms_passed)
        
        if keyY.value() == 0:
            while keyY.value() == 0: pass
            return
        if keyA.value() == 0:
            if running == True:
                offset = ms_passed
                running = False
            else:
                running = True
                start = time.ticks_ms()
                
            while keyA.value() == 0:
                pass
            time.sleep(0.001)
        time.sleep(0.001)
    
def run():
    load_in()
    global hours, minutes, seconds, stopwatch
    while True:
        if keyA.value() == 0:
            startstopwatch()
            displaytime(0)
            while keyA.value() == 0:
                pass
            time.sleep(0.05)
        if keyY.value() == 0:
            while keyY.value() == 0: pass
            return

File: cpp/apps/stopwatch.cpp
// Stopwatch — port of apps/Clock/Stopwatch.py.
#include <cstdio>

#include "pico/stdlib.h"

#include "lib/canvas.h"
#include "lib/colors.h"
#include "lib/input.h"

extern Canvas g_canvas;

namespace {

void display_time(uint32_t ms_passed) {
    uint32_t hours   = ms_passed / 3600000u;
    uint32_t minutes = (ms_passed % 3600000u) / 60000u;
    uint32_t seconds = (ms_passed % 60000u) / 1000u;
    uint32_t ms      = ms_passed % 1000u;

    char buf[16];
    snprintf(buf, sizeof(buf), "%02lu:%02lu:%02lu.%03lu",
             (unsigned long)hours, (unsigned long)minutes,
             (unsigned long)seconds, (unsigned long)ms);

    g_canvas.fill(color::BLACK);
    g_canvas.text("Stopwatch", 48, 40, color::GREEN, 2);
    g_canvas.text(buf, 24, 112, color::WHITE, 2);
    g_canvas.show();
}

uint32_t now_ms() { return to_ms_since_boot(get_absolute_time()); }

// The running/paused loop. Returns to the app's idle screen on Y.
void run_stopwatch() {
    bool running = true;
    uint32_t offset = 0;
    uint32_t start = now_ms();
    uint32_t ms_passed = 0;

    while (true) {
        input::update();
        if (running) ms_passed = (now_ms() - start) + offset;
        display_time(ms_passed);

        if (input::pressed(input::Y)) return;

        if (input::pressed(input::A)) {
            if (running) {
                offset = ms_passed;   // freeze accumulated time
                running = false;
            } else {
                start = now_ms();      // resume from the frozen offset
                running = true;
            }
        }
        sleep_ms(1);
    }
}

}  // namespace

void stopwatch_run() {
    display_time(0);
    while (true) {
        input::update();
        if (input::pressed(input::A)) {
            run_stopwatch();
            display_time(0);
        }
        if (input::pressed(input::Y)) return;
        sleep_ms(10);
    }
}
